src/libexpr/parser.y

```diff
@@ -674,7 +689,12 @@ Expr * EvalState::parseExprFromString(std::string_view s, const Path & basePath)
 Expr * EvalState::parseStdin()
 {
     //Activity act(*logger, lvlTalkative, format("parsing standard input"));
-    return parse(drainFD(0).data(), foStdin, "", absPath("."), staticBaseEnv);
+    auto buffer = drainFD(0);
+    // make sure drainFD has left some extra space for terminators
+    assert(buffer.capacity() > 2 && buffer.size() <= buffer.capacity());
```

Comment:
Why is this assertion necessary? The `+=` operator will grow the string if necessary. `buffer.size() <= buffer.capacity()` is an invariant guaranteed by `std::string`. `buffer.capacity() > 2` may fail if the input is shorter than 2 bytes.

Reply:
to avoid copying the entire input string (which the generated parser previously did under the hood). when evaluating nixpkgs those copies add up to quite a bit, but for stdin it's admittedly not as important.

Reply:
also looks like we accidentally dropped some arithmetic, it really should be
```suggestion
    assert(buffer.capacity() > 2 && buffer.size() <= buffer.capacity() - 2);
```